src/scheduler.cpp

```diff
@@ -64,13 +69,69 @@ static void generic_wait_handler(task_wait_handle wait_task)
 	thread_event.wait();
 }
 
+#if defined(EMULATE_PTHREAD_THREAD_LOCAL)
 // Wait handler function, per-thread, defaults to generic version
-static THREAD_LOCAL wait_handler thread_wait_handler = generic_wait_handler;
+static pthread_key_t thread_wait_handler_key;
+
+static void create_thread_wait_handler()
+{
+    if(pthread_getspecific(thread_wait_handler_key) == nullptr) {
+    	pthread_setspecific(thread_wait_handler_key, (void*)generic_wait_handler);
+    }
+}
+
+struct pthread_emulation_thread_wait_handler_key_initializer {
+    pthread_emulation_thread_wait_handler_key_initializer()
+    {
+        pthread_key_create(&thread_wait_handler_key, nullptr);
+    }
+        
+    ~pthread_emulation_thread_wait_handler_key_initializer()
+    {
+        pthread_key_delete(thread_wait_handler_key);
+    }
+};
+    
+static pthread_emulation_thread_wait_handler_key_initializer thread_wait_handler_key_initalizer;
+
+#else
+static THREAD_LOCAL wait_handler thread_wait_handler;
```

Comment:
For the case where there is native support for `thread_local` you can just initialize it with `static THREAD_LOCAL wait_handler thread_wait_handler = generic_wait_handler;`.

For the emulation case, you *only* need to initialize the key value in `get_thread_wait_handler` if `pthread_getspecific` returned null. Therefore you can get rid of the `create_thread_wait_handler` function.